x/bankplus/keeper/keeper.go

```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func (keeper BaseKeeper) InitializeBankPlus(ctx context.Context) {
+	keeper.loadAllInactiveAddrs(ctx)
+}
+
+// SendCoinsFromModuleToAccount transfers coins from a ModuleAccount to an AccAddress.
+// It will panic if the module account does not exist.
+func (keeper BaseKeeper) SendCoinsFromModuleToAccount(
+	ctx context.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins,
+) error {
+	senderAddr := keeper.ak.GetModuleAddress(senderModule)
+	if senderAddr.Empty() {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", senderModule))
+	}
+
+	if keeper.BlockedAddr(recipientAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", recipientAddr)
+	}
+
+	return keeper.SendCoins(ctx, senderAddr, recipientAddr, amt)
+}
+
+// SendCoinsFromModuleToModule transfers coins from a ModuleAccount to another.
+// It will panic if either module account does not exist.
+func (keeper BaseKeeper) SendCoinsFromModuleToModule(
+	ctx context.Context, senderModule, recipientModule string, amt sdk.Coins,
+) error {
+	senderAddr := keeper.ak.GetModuleAddress(senderModule)
+	if senderAddr.Empty() {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", senderModule))
+	}
+
+	recipientAcc := keeper.ak.GetModuleAccount(ctx, recipientModule)
+	if recipientAcc == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", recipientModule))
+	}
+
+	return keeper.SendCoins(ctx, senderAddr, recipientAcc.GetAddress(), amt)
+}
+
+// SendCoinsFromAccountToModule transfers coins from an AccAddress to a ModuleAccount.
+// It will panic if the module account does not exist.
+func (keeper BaseKeeper) SendCoinsFromAccountToModule(
+	ctx context.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins,
+) error {
+	recipientAcc := keeper.ak.GetModuleAccount(ctx, recipientModule)
+	if recipientAcc == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", recipientModule))
+	}
+
+	return keeper.SendCoins(ctx, senderAddr, recipientAcc.GetAddress(), amt)
+}
+
+func (keeper BaseKeeper) isInactiveAddr(addr sdk.AccAddress) bool {
+	addrString, err := keeper.addrCdc.BytesToString(addr)
+	if err != nil {
+		panic(err)
+	}
+	return keeper.inactiveAddrs[addrString]
+}
+
+// SendCoins transfers amt coins from a sending account to a receiving account.
+// This is wrapped bank the `SendKeeper` interface of `bank` module,
+// and checks if `toAddr` is a inactiveAddr managed by the module.
+func (keeper BaseKeeper) SendCoins(ctx context.Context, fromAddr, toAddr sdk.AccAddress, amt sdk.Coins) error {
+	// if toAddr is smart contract, check the status of contract.
+	if keeper.isInactiveAddr(toAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", toAddr)
+	}
+
+	return keeper.BaseSendKeeper.SendCoins(ctx, fromAddr, toAddr, amt)
```

Comment:
We can remove duplicate logic using the send restriction feature. This, or another issue should address it.